Feature/install target

guiguem opened the pull request:
This is a rather big rework of the cmake in order to include a `install` target.
The primary goal of this target is to allow clean installation where the intermediate files (like CMakeCache or .o binary files) are not in the final install folder (similarly to G4 and ROOT).
A couple of modifications were made:
- this relies on two cmake files called PackageBuilder which contain useful macros for making this easier. We could live without them and have less macros, but this made my life easier. These macros are used in other packages (HK, P-Theta...) and are rather well-maintained.
- Instead of one single library (WCSimRoot), two libraries are created:
    - WCSimRoot: related to the linkDef and root libraries
    - WCSimCore: contains all the classes needed by the WCSim executable (so that people could just link to this one instead of compiling any source file)
- the `install` target will copy all the libraries, executables, data files, cmake config files and headers to a location defined by the `-D CMAKE_INSTALL_PREFIX=XXX` option. If none is provided, make install will copy things into the binary folder (where the cmake command was run).
- A WCSimConfig.cmake file is produced: this offers the possibility to use CMake instructions like `Find(WCSim <version>)` in derived projects (very useful for HK where projects like hk-eventDisplay will use such Find instruction to find ROOT and WCSim). 
- A generic `this_wcsim.sh` file is produced with the install target, this sets PATH and LD_LIBRARY_PATH to point to this installation. (could be called within other setup bash script)
- support to Root5 was not tested, only Root6 (not sure where I can find a Root5 installation in order to test this).

guiguem committed:
Compilation and install based on PB successful

guiguem committed:
Implement last features

tdealtry commented on WCSimConfig.cmake.in:
```diff
@@ -0,0 +1,7 @@
+# DAQConfig.cmake
```
This seems like a copy/paste old comment. Best to use the new filename

tdealtry commented on this_wcsim.sh.in:
```diff
@@ -0,0 +1,6 @@
+# Sets necessary environment variables to use this installation of PTheta in ROOT
```
again, a copy/paste comment that looks odd in WCSim

tdealtry commented on this_wcsim.sh.in:
```diff
@@ -0,0 +1,6 @@
+# Sets necessary environment variables to use this installation of PTheta in ROOT
+# Also set PATH and LD_LIBRARY_PATH for convenience
+
```
This file probably sits better in either `cmake/` or `setup/`

tdealtry reviewed (changes requested):
Thanks for this!

A few minor inline comments

Another couple of things
- `rootwc` should be included in `bin/` in the output
- Could you modify `README.md` to explain how to build using the new code? Specifically
  - The requirement of CMake3 is important to list
  - One specific confusion I had was running `cmake ../WCSim && make` in the build directory doesn't create a `bin/` in the build directory - a `make install` is required

FYI I'll move to drop the GNUMakefile's and move to CMake exclusively after the hybrid/OD merge

gpronost commented:
Hi, 

I have a couple of comment:
- Can you modify ./make.sh to work with this solution?
- For ROOT5 / ROOT6 test this can be done on sukap (call `setup/env_sukap.sh 5` to load ROOT5)

setup/env_sukap.sh is actually outdated, here is the updated version (feel free to include it in the PR, I can't easily make one since I have another PR pending for the nuPRISM fork)

```
#!/bin/bash

# ROOT version as argument, default value is 6 for ROOT6
#ROOT_ARG=5
ROOT_ARG=6

if [ ! -z "$1" ] ; then
	echo "ROOT version requested to be: ${1}"
	ROOT_ARG=$1
fi

############################
# ROOT/CERN
############################

if [[ "$ROOT_ARG" == 5 ]] ; then 
	echo "Use ROOT5"
	# Note: official ROOT 5.34 installation on sukap is not compatible with cmake
	#  Temporary use Guillaume Pronost's version. Should change in the future.
	#
	export ROOT_DIR=/home/pronost/software/root-5.34.38-build
	export ROOT_STR=ROOT5
else
	echo "Use ROOT6"
	# Note: There is no official ROOT 6 installation 
	#  Temporary use Guillaume Pronost's version. Should change in the future.
	export ROOT_DIR=/home/pronost/software/root-6.22.00-build
	export ROOT_STR=ROOT6
fi
	
# Note: There is no official ROOT 6 installation 
#  Temporary use Guillaume Pronost's version. Should change in the future.
#export ROOT_DIR=/home/pronost/software/root-6.22.00-build
source ${ROOT_DIR}/bin/thisroot.sh

alias root='root -l'

############################
# GEANT4
############################

pwd=$PWD
GEANT4VERSION=geant4.10.05.p01
cd /usr/local/sklib_gcc8/geant4.10.05.p01/bin/
source geant4.sh
cd /usr/local/sklib_gcc8/geant4.10.05.p01/share/Geant4-10.5.1/geant4make/
source geant4make.sh
echo "Connect a PATH of Geant4.10"
cd "$pwd"

############################
# WCSim
############################
export BRANCH_NAME=$(git rev-parse --abbrev-ref HEAD)
export WCSIMDIR=$(pwd)-build/ROOT${ROOT_ARG}/$BRANCH_NAME
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$WCSIMDIR

```

tdealtry commented:
Thanks @gpronost

@guiguem if you do decide to pull the `env_sukap.sh` changes in, then please take them from https://github.com/WCSim/WCSim/blob/1fed529357c3bda86f8a533d52c089b5be977dbb/setup/env_sukap.sh instead (part of #336). There's a couple of quality of life improvements (use the variable `$GEANT4VERSION`, checking that the `$ROOT_ARG` variable is valid (i.e. either 5 or 6, setup an environment vairable to point to the github source directory)

Anyway, checking `make.sh` works (and updating it if it doesn't) is a very good idea. This file should be resource agnostic (i.e. can be run anywhere, not just sukap)

guiguem committed:
Implement @gpronost and @tdealtry comments

guiguem commented:
Hi @gpronost and @tdealtry sorry for the slow ping!
Thank you both for the suggestions.
- I have fixed the typos (and not correctly copied files)
- I made a Dockerfile.cmake file which installs using this new system; I am not sure whether I should replace the Dockerfile that already exists.
- I just updated the make.sh to use cmake3 and the new build/install system. I was able to test it in a docker container.
- I didn't touch the env_sukap.sh since I don't have access to Sukap (not a SK member).

Please let me know if I forgot something!

Some missing items I can think of:
- we would need to have cmake3 in the new base container. I can for sure update the Dockerfile.base, but I am not sure I have the proper privileges for pushing a new image.
- I am rather unclear about the whole Continuous Integration: it seems to use make and not cmake. So maybe we can just let it this way and once you merge the hybrid/OD thing, we can revisit this?

tdealtry commented:
Thanks for this @guiguem!

- I think it's a good idea to keep the 2 versions of Dockerfile for now, just to keep the CI happy. As you say, can revisit it later (once OD/hybrid merge successful)
- @brichards64 is the DockerHub owner for WCSim. I'll merge this in now (with cmake3 install in Dockerfile.cmake), but we would like a new base building/pushing with cmake3 included when possible (I'll put cmake3 in the Dockerfile.base install list in another PR to make it easier)
